sdlib/d/gc/threadpool.d

```diff
@@ -0,0 +1,273 @@
+module d.gc.threadpool;
+
+struct ThreadPool {
+private:
+	import d.sync.mutex;
+	Mutex mutex;
+
+	// How many threads are executing work.
+	uint activeThreads;
+	// Stop processing work, exit the thread.
+	bool exitFlag;
+
+	import core.stdc.pthread;
+	pthread_t[] threads;
+
+	// Current work to do.
+	void delegate(uint) work;
+	// How many threads should execute the work
+	uint scheduled;
+
+	static void* threadEntry(void* ctx) {
+		// Detach this thread -- it should not be scanned or stopped.
+		import d.gc.thread;
+		detachSelf();
+		auto pool = cast(shared(ThreadPool*)) ctx;
+		pool.executeWork(false);
+		return null;
+	}
+
+public:
+
+	/**
+	 * Start threads using the given thread buffer. All threads will wait
+	 * for work dispatched to them using the `dispatch` function. Optimal
+	 * thread buffer size should be number of cores, or number of cores - 1
+	 * if you intend to use the calling thread for work.
+	 *
+	 * Note that these threads will be marked as detached, so they will not
+	 * be scanned if a GC cycle is run.
+	 */
+	void startThreads(pthread_t[] threadBuffer) shared {
+		mutex.lock();
+		scope(exit) mutex.unlock();
+
+		(cast(ThreadPool*) &this).startThreadsImpl(threadBuffer);
+	}
+
+	/**
+	 * Dispatch work to the threads. `count` instances of the `work`
+	 * delegate are run, with each call to the delegate receiving an index
+	 * from `0` to `count - 1`. If `count` is greater than the number of
+	 * threads in the pool, threads will re-call the delegate, until
+	 * `count` calls to the delegate are made. No ordering of the index is
+	 * guaranteed, but it is guaranteed that only one instance of an index
+	 * will be used.
+	 */
+	void dispatch(void delegate(uint) work, uint count) shared {
+		mutex.lock();
+		scope(exit) mutex.unlock();
+
+		(cast(ThreadPool*) &this).dispatchImpl(work, count);
+	}
+
+	/**
+	 * Wait for all work to be complete. `useThisThread` set to true will
+	 * use the calling thread to also do work if needed.
+	 */
+	void waitForIdle(bool useThisThread = true) shared {
```

Comment:
> First there is a bit of featuritis, as why wouldn't we want to get more work done faster?

So in the scanner, there is a difference in running the scan function from the main thread, and running it from the worker threads. Namely, in the worker threads, the GC is turned off, whereas in the main thread it is not turned off. Although I moved this code to the worker delegate anyway, since turning off the GC is harmless at this point (it has already been turned off), it got me thinking, you might want to do some things in the worker threads that would be incorrect for the main thread. So I added this flag as an option.

I can remove this feature for now, and we can add it when we need it.

> But second, we assume no more work will be added once we exhaust the current stash of work, but we have no way to know that, and if that happens, we'll be stuck waiting again instead of using this thread.

Work will not be added because the main thread is the only one who can add work. If you tried to dispatch more tasks while in one of the worker threads, I think there could be a race condition.

Note that this mechanism is not necessarily how the threads will divide the actual work (e.g. the scanning of a specific range). This abstraction is for creating threads, and having them run a specific task. Anything involving a changing list of work should be handled externally. There is no allocating inside here. The goal of this is to manage threads specifically for the GC, which is a restrained environment.